keps/sig-storage/4979-evented-desired-state-of-world-populator/README.md

```diff
@@ -0,0 +1,399 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes optimizing the loop iteration period (currently fixed at 100ms) in the Desired State of the World Populator (DSWP). The enhancement involves dynamically increasing the sleep period when no changes are detected and reacting to gRPC event streams from the CRI implementation to reduce unnecessary processing.
+
+## Motivation
+
+In the volume manager, the Desired State of the World executes a populator loop every 100ms, regardless of whether any changes have occurred. This fixed frequency may result in unnecessary CPU cycles during idle periods. By adopting an event-based approach, the kubelet can respond precisely when changes occur, improving performance and reducing system overhead.
+
+### Goals
+
+1. Dynamically adjust the populator loop interval based on system activity.
+2. Respond promptly to events, ensuring up-to-date DSWP cache.
+3. Maintain existing functionality as a fallback to ensure reliability.
+
+### Non-Goals
+ 
+1. Completely remove the batch loop period.
+2. Change the existing DSWP logic.
+
+## Proposal
+
+The Desired State of the World Populator will listen to gRPC event streams from the CRI implementation. Specifically, the CONTAINER_CREATED_EVENT and CONTAINER_DELETED_EVENT will trigger the populator loop.
+During periods of inactivity, the populator loop interval will increase by 100ms increments after the third execution, up to a maximum of 1 second. If an event is detected, the interval resets to the default 100ms. This approach ensures responsiveness while reducing CPU usage.
```

Comment:
DSWP also triggers online expansion of volumes in kubelet if requested by the user. How does that work? Does this mean, online resizing will still work, but it will be slower?

Reply:
I agree with you. 

The KEP will improve performance for mount/unmount operations. However, for online volume resize, users may experience some delay (100 ms currently VS <= 1 second).

My considerations on this:
1. Ship the current KEP with this behavior: the resize request delay will be documented and highlighted when using this KEP.
2. Reduce the maximum sleep period to 500 ms or keep it ≤ 1 second?
3. Catching volume resize event with informer ? (a good choice ? coupling kubelet  with non node-component  ? ) 

What do you think about the fact that this delay in resizing is a reasonable trade-off for the overall improvement in mount/unmount performance?

Reply:
based on the last changes (event based on pod manager state) , I suppose the online resizing will be catched as  a pod update (no delay exepected) .